crates/executor/src/blockchain_tree/chain.rs

```diff
@@ -0,0 +1,258 @@
+    }
+
+    /// Create new chain that joins canonical block
+    /// If parent block is the tip mark chain fork.
+    pub fn new_canonical_fork<PROVIDER: StateProvider, CONSENSUS: Consensus>(
+        block: &SealedBlockWithSenders,
+        parent_header: SealedHeader,
+        canonical_block_hashes: &BTreeMap<BlockNumber, BlockHash>,
+        chain_spec: Arc<ChainSpec>,
+        provider: &PROVIDER,
+        consensus: &CONSENSUS,
+    ) -> Result<Self, Error> {
+        // verify block against the parent
+        consensus.validate_header(block, U256::ZERO)?;
+        consensus.pre_validate_header(block, &parent_header)?;
+        consensus.pre_validate_block(block)?;
+
+        // substate
+        let mut substate = SubStateData::default();
+        let empty = BTreeMap::new();
+
+        let unseal = block.clone().into_components().0.unseal();
+        let provider =
+            SubStateWithProvider::new(&mut substate, provider, &empty, canonical_block_hashes);
+        let mut state_provider = SubState::new(State::new(provider));
+
+        // execute block
+        let changesets = vec![Executor::new(chain_spec, &mut state_provider)
+            .execute_and_verify_receipt(&unseal, U256::MAX, None)?];
+
+        Ok(Self { substate, changesets, blocks: BTreeMap::from([(block.number, block.clone())]) })
+    }
+
+    /// DONE
+    /// Create new chain that branches out from existing side chain.
+    pub fn new_chain_fork<PROVIDER: StateProvider, CONSENSUS: Consensus>(
+        &self,
+        block: SealedBlockWithSenders,
+        side_chain_block_hashes: BTreeMap<BlockNumber, BlockHash>,
+        canonical_block_hashes: &BTreeMap<BlockNumber, BlockHash>,
+        chain_spec: Arc<ChainSpec>,
+        provider: &PROVIDER,
+        consensus: &CONSENSUS,
+    ) -> Result<Self, Error> {
+        let parent_nubmer = block.number - 1;
+        let parent = self
+            .blocks
+            .get(&parent_nubmer)
+            .ok_or(ExecError::BlockNumberNotFoundInChain { block_number: parent_nubmer })?;
+
+        // verify block against the parent
+        consensus.validate_header(&block, U256::ZERO)?;
+        consensus.pre_validate_header(&block, parent)?;
+        consensus.pre_validate_block(&block)?;
+
+        // revert changesets
+        let revert_from = self.changesets.len() - (self.tip().number - parent.number) as usize;
+        let mut substate = self.substate.clone();
+
+        // Revert changesets to get the state of the parent that we need to apply the change.
+        substate.revert(&self.changesets[revert_from..]);
+
+        let provider = SubStateWithProvider::new(
+            &mut substate,
+            provider,
+            &side_chain_block_hashes,
+            canonical_block_hashes,
+        );
+
+        let unseal = block.clone().into_components().0.unseal();
+        // Create state provider with cached state
+        let mut state_provider = SubState::new(State::new(provider));
```

Comment:
substate wrapping seems a bit convoluted here and seems to happen quite often in the chain

Reply:
Agreed:
https://github.com/paradigmxyz/reth/commit/0978d9a000d3fc0ffe2d0c7ab806a52a846bd41c#diff-03e693dceaec1e158ab3f33250a3668cb1dd3ed35d56c45adf0e23309cf08873R31-R38

Reply:
ExecutorFactory hides substate wrapping